dcp_prototype/backend/wrangling/migrations/scripts/metadata_aggregator.py

```diff
@@ -0,0 +1,166 @@
+            s3_client.download_file(bucket, object_name, download_dest)
+        except Exception as e:
+            print(f"Failed {str(e)} : {bucket} {object_name} {download_dest}")
+
+        filequeue.task_done()
+
+
+def download_files(s3_uri, num_threads, dirname):
+
+    bucket_name = urlparse(s3_uri).netloc
+    prefix = urlparse(s3_uri).path
+    if prefix.startswith("/"):
+        prefix = prefix[1:]
+
+    client = boto3.client("s3")
+    paginator = client.get_paginator("list_objects")
+    page_iterator = paginator.paginate(Bucket=bucket_name, Prefix=prefix)
+
+    filequeue = queue.Queue()
+    nfiles = 0
+    prefixparts = len(prefix.split("/"))
+    for page in page_iterator:
+        bucket_objects = page.get("Contents")
+        for object in bucket_objects:
+            object_filename = object.get("Key")
+            if "data_files" in object_filename:
+                continue
+            parts = object_filename.split("/")
+            if len(parts) == prefixparts + 1:
+                filequeue.put(parts[-1])
+                nfiles += 1
+                if nfiles % 1000 == 0:
+                    print(nfiles, end=" ", flush=True)
+
+    print(nfiles)
+    threads = []
+    for _ in range(num_threads):
+        thread = threading.Thread(target=download_file, args=(prefix, bucket_name, filequeue, dirname))
+        thread.start()
+        threads.append(thread)
+
+    filequeue.join()
+    for _ in range(num_threads):
+        filequeue.put(None)
+    for thread in threads:
+        thread.join()
+
+    print("downloaded nfiles=", nfiles)
+
+
+def create_tarfile(dirname):
+    tarfilename = dirname + ".tar.gz"
+    nfiles = 0
+    with tarfile.open(tarfilename, "w:gz") as tar_handle:
+        for filename in os.listdir(dirname):
+            tar_handle.add(os.path.join(dirname, filename))
+            nfiles += 1
+
+    print("files in tar", nfiles)
+    return tarfilename
+
+
+def upload_to_s3(src_file, dest_s3_uri):
+    bucket = urlparse(dest_s3_uri).netloc
+    object_name = urlparse(dest_s3_uri).path
+    if object_name.startswith("/"):
+        object_name = object_name[1:]
+
+    s3_client = boto3.client("s3")
+    try:
+        print(f"Upload: {src_file}, {bucket}, {object_name}")
+        s3_client.upload_file(src_file, bucket, object_name)
+        print(f"Upload complete: {object_name}")
+    except ClientError as e:
+        print("Error:", str(e))
+        return False
+
+    return True
+
+
+def aggregate(input_directory, num_threads, dirname, output_object):
+    if dirname.endswith("/"):
+        dirname = dirname[:-1]
+    download_files(input_directory, num_threads, dirname)
+    tarname = create_tarfile(dirname)
+    if output_object:
+        return upload_to_s3(tarname, output_object)
+    return True
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "-i", "--input_directory", help="A data directory contains metadata files that were a part of a single DCP 1.0"
```

Comment:
data directory or s3 bucket?

Reply:
Actually it can be a local directory, an s3 bucket, a tar.gz file (either local or on S3).  I'll update the help

Reply:
Yes, this is a location in S3.  I'll clarify the help doc.